specification/dartLangSpec.tex

```diff
@@ -15861,7 +15863,11 @@ \subsection{Lexical Lookup}
 
 \LMHash{}%
 A lexical lookup yields a result which is
-a declaration, an import prefix, or a member signature.
+a declaration, an import prefix,
+or the special value
+\Index{\PrependThisValue},
```

Comment:
Why signal this with a special value instead of just saying that "lexical lookup fails"?  That seems much more straightforward.  Either lexical lookup succeeds and yields a declaration, or it fails and we try to prepend `this`?

Reply:
That's tempting, but it's not obvious to me that it would be a simplification.

The idea is that we handle the compile-time errors in one location (the case list starting at line 15941), such that there will not be any compile-time errors caused by a lookup failure at the locations where lexical lookup is used. (We never specify propagation of errors, so when the "client" says `perform a lexical lookup for the name $n$` we proceed with the outcome and do not mention "but we drop any further processing of this expression if an error occurred").

So the good thing about this organization is that we mention all the errors of lexical lookups in one piece of text.

I think we do need to mention at least some errors here, and I think we'd have to repeat the same errors "at the client location" if we just report a failure of the lookup, covering all the real errors plus the situations where we will prepend `this` and try again.

We have 50 occurrences of `lexical lookup`, 30 of them outside the `Lexical Lookup` section, so it is likely to cause a substantial duplication of each of these phrases specifying when the failure is actually an error.

We could try to do it, it is possible that we can rule out some of the errors for some occurrences of `lexical lookup` outside the `Lexical Lookup` section.

Reply:
I'm not sure I understand.  Currently, you claim that "lexical lookup cannot fail".  But of course it can, you just mean that either lexical lookup causes an error, or it produces something, and that something can be either a declaration or a special value saying "lexical lookup failed to find something".    I'm saying, why not just get rid of the special value and say that one of three things happen:
  - lexical lookup causes an error (as it may now)
  - lexical lookup finds a declaration (as it may now)
  - lexical lookup fails to find a declaration (what you're calling PrependThisValue)

This seems exactly isomorphic, but avoids indirecting through an additional definition (PrependThisValue)

Reply:
> We could say that lexical lookup failed to find a declaration, but that doesn't match the situation precisely: lots of situations where we fail to find a declaration are an error, but PrependThisValue implies that a number of error checks have been performed and a suitable declaration _does_ exist.

I don't really agree with this, at least if I'm understanding what's here.  I think saying that lexical lookup failed does match the situation precisely.  This algorithm either finds a single declaration in the lexical scope; or it emits an error because there is no possible suitable thing in either the lexical or the instance scope; or it says "there's nothing in the lexical scope, but there's something that might (or might not) work in the instance scope, go do some additional work".  The last is what you are calling "PrependThisValue".  It still seems much cleaner to me to just call these "Found a declaration", "Caused an error", and "Didn't find a declaration".

Reply:
> So the 'caller' doesn't have to take into account that the lexical lookup failed to find anything. When it actually fails to find anything, a compile-time error has occurred.

Nothing I'm suggesting changes this.

Reply:
> because 'fails to find a declaration' would presumably also cover at least some of the many error case

I'm not proposing that.  Why would it presumably also cover that?

Reply:
> because 'fails to find a declaration' would presumably also cover at least some of the many error cases, and PrependThisValue is only yielded when we have actually found a member with the requested name and established that prepending `this` is appropriate.

If you prefer, you could say "fails to return a declaration"?  But I actually think "fails to find a declaration" is just fine.  It found something that might (or might not) be the right thing, but it's not returning anything because it doesn't know whether there is an unambiguous answer.

Reply:
> As I mentioned, we're using lexical lookup in about 30 locations, and I don't think it's helpful to repeat that many times which errors we should raise when nothing was found (or we found a getter and wanted a setter, etc.).

This is part of what I'm not getting entirely.  I agree that you probably need to catch the getter/setter mismatch here (are we sure we're specifying this as implemented, BTW?).  But the case where it's not found I think is already handled by the logic for `this.id`, so why catch it here?

Reply:
> This algorithm either finds a single declaration in the lexical scope; or it emits an
> error because there is no possible suitable thing in either the lexical or the instance
> scope; or it says "there's nothing in the lexical scope, but there's something that
> might (or might not) work in the instance scope, go do some additional work"

>  It still seems much cleaner to me to just call these "Found a declaration",
> "Caused an error", and "Didn't find a declaration".

I don't think we should start specifying propagation of errors, we otherwise never do that. So if the lexical lookup encountered an error then we just silently ignore it at the 'call site', exactly like we'd ignore a type error in `b` when we specify `if (b) S`: we require that `b` has a type assignable to `bool`, but we never mention what to do if `b` has no type because it isn't type correct, so it's like a `throw` that we don't catch at the call site. So I wouldn't consider "Caused an error" to be the result of a lexical lookup in a way that is similar to "Found a declaration": The call site needs to talk about the latter, but would not mention the former.

"Found a declaration" matches the situation where the lexical lookup yields a declaration, so that's fine. We're currently handling import prefixes separately, but that's just because it seems somewhat odd to consider an import directive as a declaration of an import prefix, especially because we can have the same prefix on multiple imports.

"Didn't find a declaration" matches the situation where there was no declaration with the requested basename at all (enclosing lexical scopes + interface of `this` + extension methods), but that gives rise to an error (so, again, we ignore it at the call site).

The remaining situations are those where we _have_ found a declaration with the requested name in the interface of `this`, or `this` has no such basename and we found one in an extension. Is the best possible name for this situation really "Didn't find a declaration"?

Reply:
> Is the best possible name for this situation really "Didn't find a declaration"?

Yes.  Because the algorithm you are specifying is **lexical** lookup.  Not **lexical and instance** lookup.  I get that you want to do some error checking in this algorithm, and that we just silently propagate error cases.  But if we are doing lexical lookup, and we ignore the error cases, then either it found something in the lexical scope or it didn't.  If it didn't, we lookup in the instance scope.  To use your terminology, you are specifying this as returning an element of `datatype LexicalDeclaration of declaration | PrependThis`.  I'm saying that that datatype is precisely isomorphic to `Declaration option` so why go to the trouble to introduce a new datatype instead of just using the `option` datatype?  The new datatype is just one more thing for the reader to load into their head.